src/services/saml-config-service.js

```diff
@@ -0,0 +1,79 @@
+import xpath from 'xpath';
+import xmldom from 'xmldom';
+import memoizee from 'memoizee';
+import HttpClient from '../api-clients/http-client.js';
+import ServerConfig from '../bootstrap/server-config.js';
+
+const { DOMParser } = xmldom;
+
+const REQUEST_TIMEOUT_IN_MS = 30 * 1000;
+const REDIRECT_BINDING = 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect';
+
+const xpathSelect = xpath.useNamespaces({
+  sm: 'urn:oasis:names:tc:SAML:2.0:metadata',
+  ds: 'http://www.w3.org/2000/09/xmldsig#'
+});
+
+export default class SamlConfigService {
+  static get inject() { return [ServerConfig, HttpClient]; }
+
+  constructor(serverConfig, httpClient) {
+    this.httpClient = httpClient;
+    this.serverConfig = serverConfig;
+    this._identityProviderMap = new Map();
+  }
+
+  getIdentityProviders() {
+    return [...this._identityProviderMap.values()];
+  }
+
+  getIdentityProviderByKey(providerKey) {
+    return this._identityProviderMap.get(providerKey) || null;
+  }
+
+  async loadXmlDoc(url) {
+    try {
+      const response = await this.httpClient.get(url, { responseType: 'text', timeout: REQUEST_TIMEOUT_IN_MS });
+      return new DOMParser().parseFromString(response.data);
+    } catch (error) {
+      if (error.code === 'ECONNABORTED') {
+        throw new Error(`Timeout trying to fetch "${url}"`);
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  resolveMetadata(xmlDoc, entityId) {
```

Comment:
Maybe we should wrap `resolveMetadata` and `resolveIdentityProviderMetadata` in `try`-`catch` and gracefully handle the mishap.

Reply:
I deliberately didn't as we mostly do not handle errors in the services. They "bubble up" (in this case the scheduled task will handle it and not update the config and just log the error).

Reply:
Of course, I see.